zarr/storage.py

```diff
@@ -544,6 +544,8 @@ def __getitem__(self, item):
     def __setitem__(self, item, value):
         with self.write_mutex:
             parent, key = self._require_parent(item)
+            if not isinstance(value, self.cls):
+                value = ensure_bytes(value)
```

Comment:
I wonder, should this just be:

```python
    def __setitem__(self, item, value):
        with self.write_mutex:
            parent, key = self._require_parent(item)
            value = ensure_bytes(value)
            parent[key] = value
```

I.e., we only allow setting of bytes values, we don't allow any fiddling with the internal hierarchy of dicts?